keps/draft-repo.md

```diff
@@ -0,0 +1,108 @@
+
+* The structure of a repository
+* Defining authoritative repositories: incubator & graduated
+* Working with multiple repositories
+* Packaging and Repository management tooling
+* Configuration of repositories for the CLI
+* CLI Tooling for search, packaging, and pushing to/from repository
+* Define the rules and means to promote incubated operators to the graduated repository.
+
+By providing a repository (or set of repositories), KUDO will provide a mechanism for administrators to easy discover and install KUDO operators.  [Helm has provided some prior art](https://github.com/helm/helm/blob/master/docs/chart_repository.md) which we should align closely to from a mental model stand point.
+
+## Motivation
+
+We need a way for the operator author to update a repository with his/her operator.  We also need a way for the operator to promote an operator as being production ready. Finally we need a KUDO user to point to different repositories.  
+
+### Goals
+
+- Automate the creation of a operator tarball
+- Automate the creation or updating of the repository index with a new operator tarball
+- Define rules for incubator and graduated along with graduation tooling.
+- Define how an admin can create their own repository
+- Install operator from incubator instead of graduated repository
+
+### Non-Goals
+
+- Manage repos of non-KUDO applications (e.g. Helm Charts)
+- External formats defined by [KEP0013](keps/0013-external-specs.md)
+- Repository cacheing
+
+## Proposal
+
+### Create Operator Tarball
+
+An operator package / bundle as defined in [KEP0010](keps/0010-package-manager.md).  The storage package at the repository is a tarball. We need a way to create this tarball in a uniformed way. KUDO shall a way to generate a tarball based on the standard file system layout.  As an example: `kubectl kudo package docs/examples/zookeeper/`.  This will create a `zookeeper-3.4.10.tgz` based on parsing the `operator.yaml` for version details. It would be best if this included linting to ensure that the static structure of the operator is correct. At this point, it will guarantee that operator folder has `operator.yaml` and `params.yaml` along with a templates directory.  It will create the following operator tarball `{operator.yaml:name}-{operator.yaml:version}.tgz`.
+
+### Create Repository Index
+
+KUDO needs the ability to create an index file for the repository. Something like `kubectl kudo repo index new-repo --url https://kudo-repo.storage.googleapis.com`. In this example, the `kudo repo index` is the command. The `new-repo` is a folder containing operator tarballs. The index file defined in [KEP0010](keps/0010-package-manager.md) will be created using the url provided for links in the file.
```

Comment:
Could the URLs be relative instead of absolute?

Reply:
seems like an implementation detail that I'm not entire sure about at this point.  This kep is to capture the need to create a repo index.   if we can find ways to make life easier 👍.   The helm docs used as inspiration does not.